### PR TITLE
Provide support to OAuth2 authentication for mailTo transport Listener

### PR DESCRIPTION
**Pull request by dulanjalidilmi**
## Purpose
Fix https://github.com/wso2/api-manager/issues/1029

**Introduced parameters are as follows.**
Parameter | Description | Possible Values | Default Value | Required
-- | -- | -- | -- | --
transport.mail.auth.mechanism | basic authentication or oauth authentication | basic, oauth | basic | Default is set to basic.
transport.mail.oauth.grantType | grant type | authorization_code, client_credentials | - | Required if the auth mechanism is oauth
transport.mail.oauth.clientId | client id | - | - | Required if the auth mechanism is oauth
transport.mail.oauth.clientSecret | client secret | - | - | Required if the auth mechanism is oauth
transport.mail.oauth.tokenUrl | token endpoint url | - | - | Required if the auth mechanism is oauth
transport.mail.oauth.scope | scope | - | - | Required if the grant type is client_credentials
transport.mail.oauth.refreshToken | refresh token generated | - | - | Required if the grant type is authorization_code
transport.mail.oauth.accessToken | access token generated | - | - | Optional

### CODE REVIEW

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthHandler.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        try {
+            Store store = session.getStore(protocol);
+            try {
+                if (log.isDebugEnabled()) {
+                    log.debug("Connecting to mail service.");
+                }
+                store.connect(username, getToken());
+                return store;
+            } catch (Exception e) {
+                log.error("An error occurred while trying to connect to mail server.");
```
Comment (arunans23) [inline modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthHandler.java:63]: Shall we include some information related to the mail server?

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java [hunk 2]
```diff
@@ -105,22 +105,18 @@ private void checkMail(final PollTableEntry entry, InternetAddress emailAddress)
             try {
                 retryCount++;
                 if (log.isDebugEnabled()) {
-                    log.debug("Attempting to connect to POP3/IMAP server for : " +
-                        entry.getEmailAddress() + " using " + session.getProperties());
+                    log.debug("Attempting to connect to POP3/IMAP server for : " + entry.getEmailAddress());
```
Comment (dulanjalidilmi) [inline modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java:108]: Let's print more properties here.

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthHandler.java (hunk 1 above)
Comment (dulanjalidilmi) [reply modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthHandler.java]: Fixed.

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java (hunk 2 above)
Comment (dulanjalidilmi) [reply modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java]: Fixed.

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java [hunk 3]
```diff
@@ -96,31 +98,30 @@ private void checkMail(final PollTableEntry entry, InternetAddress emailAddress)
...
+                            entry.getEmailAddress());
                 }
 
+                store = AuthUtils.connect(entry);
+
                 // were we able to connect?
                 connected = store.isConnected();
 
                 if (connected) {
+                    if (log.isDebugEnabled()) {
```
Comment (SanojPunchihewa) [inline modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java:122]: Let's remove this `log.isDebugEnabled()` here as well.

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/PollTableEntry.java [hunk 4]
```diff
@@ -240,6 +272,31 @@ public boolean loadConfiguration(ParameterInclude paramIncl) throws AxisFault {
...
+                oAuthConfig.setScope(mailTransportOAuthProperties.get(OAuthConstants.TRANSPORT_MAIL_OAUTH_SCOPE));
+                oAuthConfig.setTokenUrl(
+                        mailTransportOAuthProperties.get(OAuthConstants.TRANSPORT_MAIL_OAUTH_TOKEN_URL));
+                oAuthConfig.setTokenId(UIDGenerator.generateUID());
+                setOAuthConfig(oAuthConfig);
+                if (StringUtils.isNotBlank(oAuthConfig.getAccessToken())) {
+                    // this is a non-expiring token
+                    updateTokenCache();
+                }
             }
```
Comment (arunans23) [inline modules/mail/src/main/java/org/apache/axis2/transport/mail/PollTableEntry.java:300]: Shall we move this to a different method to keep this function shorter?

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthConstants.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    public static final String PARAM_REFRESH_TOKEN = "&refresh_token=";
+    public static final String SCOPE = "&scope=";
+    public static final String AMPERSAND = "&";
+    public static final String EQUAL_MARK = "=";
+
+    // elements in the oauth response
+    public static final String ACCESS_TOKEN = "access_token";
+    public static final String EXPIRES_IN = "expires_in";
+
+    public static final String EMPTY_STRING = "";
```
Comment (arunans23) [inline modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthConstants.java:54]: Shall we use StringUtils.EMPTY?

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/ClientCredentialsHandler.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * This class is used to handle Client Credentials grant oauth.
+ */
+public class ClientCredentialsHandler extends OAuthHandler {
+
+    private static final Log log = LogFactory.getLog(ClientCredentialsHandler.class);
+    private final String scope;
```
Comment (arunans23) [inline modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/ClientCredentialsHandler.java:29]: Use all caps for final variables.

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthConstants.java (hunk 5 above)
Comment (SanojPunchihewa) [reply modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthConstants.java:54]: It seems we are not using this constant

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java (hunk 3 above)
Comment (dulanjalidilmi) [reply modules/mail/src/main/java/org/apache/axis2/transport/mail/MailTransportListener.java]: Fixed.

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/PollTableEntry.java (hunk 4 above)
Comment (dulanjalidilmi) [reply modules/mail/src/main/java/org/apache/axis2/transport/mail/PollTableEntry.java]: Created a new method "updateOAuthConfig"

File: modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthConstants.java (hunk 5 above)
Comment (dulanjalidilmi) [reply modules/mail/src/main/java/org/apache/axis2/transport/mail/auth/oauth/OAuthConstants.java]: Removed EMPTY_STRING;

**Review by arunans23 — DISMISSED** [review]

**Review by arunans23 — APPROVED** [review]

**Pull request merged**